lib/cerner/oauth1a/signature.rb

```diff
@@ -0,0 +1,158 @@
+# frozen_string_literal: true
+
+require 'base64'
+require 'cerner/oauth1a/protocol'
+require 'openssl'
+require 'uri'
+
+module Cerner
+  module OAuth1a
+    # Public: OAuth 1.0a signature utilities.
+    module Signature
+      METHODS = ['PLAINTEXT', 'HMAC-SHA1'].freeze
+
+      # Public: Creates a PLAINTEXT signature.
+      #
+      # Reference: https://tools.ietf.org/html/rfc5849#section-3.4.4
+      #
+      # keywords - The keyword arguments:
+      #            :client_shared_secret - Either the Accessor Secret or the Consumer Secret.
+      #            :token_shared_secret  - The Token Secret.
+      #
+      # Returns a String containing the signature.
+      def self.sign_via_plaintext(client_shared_secret:, token_shared_secret:)
+        client_shared_secret = Protocol.percent_encode(client_shared_secret)
+        token_shared_secret = Protocol.percent_encode(token_shared_secret)
+        "#{client_shared_secret}&#{token_shared_secret}"
+      end
+
+      # Public: Creates a HMAC-SHA1 signature.
+      #
+      # Reference: https://tools.ietf.org/html/rfc5849#section-3.4.2
+      #
+      # keywords - The keyword arguments:
+      #            :client_shared_secret  - Either the Accessor Secret or the Consumer Secret.
+      #            :token_shared_secret   - The Token Secret.
+      #            :signature_base_string - The Signature Base String to sign. See
+      #                                     Signature.build_signature_base_string.
+      #
+      # Returns a String containing the signature.
+      def self.sign_via_hmacsha1(client_shared_secret:, token_shared_secret:, signature_base_string:)
+        client_shared_secret = Protocol.percent_encode(client_shared_secret)
+        token_shared_secret = Protocol.percent_encode(token_shared_secret)
+        signature_key = "#{client_shared_secret}&#{token_shared_secret}"
+        signature = OpenSSL::HMAC.digest('sha1', signature_key, signature_base_string)
+        encoded_signature = Base64.encode64(signature)
```

Comment:
`Base64.strict_encode64` won't add the extra new-line: https://ruby-doc.org/stdlib-2.6.5/libdoc/base64/rdoc/Base64.html#method-i-strict_encode64

Reply:
The OAuth 1.0 specification expects RFC 2045, which is what 'encode64' does. The 'strict_encode64' uses RFC 4648. I haven't done the specific testing, but I'm not sure yet that the two are completely compatible. I know that that urlsafe/filesafe one is not compatible.

Reply:
I think you're correct. The underlying implementation just goes to the pack method and uses 'm' or 'm0' for the two encoding method. That documentation is as follows ...

```
m            | String  | base64 encoded string (see RFC 2045, count is width)
             |         | (if count is 0, no line feed are added, see RFC 4648)
```

Reply:
Switch to strict_encode64 - 946709e